File: java-apts/Dson-Apt/src/main/java/cn/wjybxx/dsonapt/AptFieldProps.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dsonapt;

import cn.wjybxx.apt.AptUtils;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.AnnotationValue;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Types;
import java.util.List;
import java.util.Map;

/**
 * 我们将字段的所有信息都收集该类上，这样可以更好的支持LinkerBean。
 * 因此该类关联的几个注解的数据，而不仅仅是{@code FieldProps}。
 *
 * @author wjybxx
 * date 2023/4/6
 */
class AptFieldProps {

    public static final String TYPE_END_OF_OBJECT = "END_OF_OBJECT";
    public static final String TYPE_BINARY = "BINARY";
    public static final String TYPE_EXT_STRING = "EXT_STRING";
    public static final String TYPE_EXT_INT32 = "EXT_INT32";
    public static final String TYPE_EXT_INT64 = "EXT_INT64";
    public static final String TYPE_EXT_DOUBLE = "EXT_DOUBLE";

    public static final String DEFAULT_NUMBER_STYLE = "SIMPLE";
    public static final String DEFAULT_STRING_STYLE = "AUTO";

    /** 字段序列化时的名字 */
    public String name = "";
    /** 取值方法 */
    public String getter = "";
    /** 赋值方法 */
    public String setter = "";

    /** 实现类 */
    public TypeMirror implMirror;
    /** 写代理方法名 */
    public String writeProxy = "";
    /** 读代理方法名 */
    public String readProxy = "";

    /** 绑定类型 */
    public String dsonType = null; // 该属性只有显式声明才有效
    public int dsonSubType = 0;

    /** 绑定style */
    public String numberStyle = DEFAULT_NUMBER_STYLE;
    public String stringStyle = DEFAULT_STRING_STYLE;
    public String objectStyle = null; // 该属性只有显式声明才有效

    /** 序列化是否忽略 -- null表示未指定 */
    public Boolean ignore;

    // region parse
    @Nonnull
    public static AptFieldProps parse(Types typeUtils, VariableElement element, TypeMirror annoMirror) {
        final AnnotationMirror annotationMirror = AptUtils.findAnnotation(typeUtils, element, annoMirror);
        if (annotationMirror == null) {
            return new AptFieldProps();
        }

        final AptFieldProps props = new AptFieldProps();
        final Map<String, AnnotationValue> annoValueMap = AptUtils.getAnnotationValuesMap(annotationMirror);
        props.name = getStringValue(annoValueMap, "name", props.name);
        props.getter = getStringValue(annoValueMap, "getter", props.getter);
        props.setter = getStringValue(annoValueMap, "setter", props.setter);

        props.dsonType = getEnumConstantName(annoValueMap, "dsonType", null);
        props.dsonSubType = getIntValue(annoValueMap, "dsonSubType", props.dsonSubType);

        props.numberStyle = getEnumConstantName(annoValueMap, "numberStyle", props.numberStyle);
        props.stringStyle = getEnumConstantName(annoValueMap, "stringStyle", props.stringStyle);
        props.objectStyle = getEnumConstantName(annoValueMap, "objectStyle", props.objectStyle);

        props.writeProxy = getStringValue(annoValueMap, "writeProxy", props.writeProxy);
        props.readProxy = getStringValue(annoValueMap, "readProxy", props.readProxy);
        final AnnotationValue impl = annoValueMap.get("impl");
        if (impl != null) {
            props.implMirror = AptUtils.getAnnotationValueTypeMirror(impl);
        }
        return props;
    }

    public void parseIgnore(Types typeUtils, VariableElement element, TypeMirror ignoreTypeMirror) {
        AnnotationMirror annotationMirror = AptUtils.findAnnotation(typeUtils, element, ignoreTypeMirror);
        if (annotationMirror != null) {
            this.ignore = AptUtils.getAnnotationValueValue(annotationMirror, "value");
        }
    }

    private static String getStringValue(Map<String, AnnotationValue> annoValueMap, String pname, String def) {
        AnnotationValue annoValue = annoValueMap.get(pname);
        if (annoValue == null) return def;
        String str = (String) annoValue.getValue();
        return str.trim();
    }

    private static Integer getIntValue(Map<String, AnnotationValue> annoValueMap, String pname, Integer def) {
        AnnotationValue annoValue = annoValueMap.get(pname);
        if (annoValue == null) return def;
        return (Integer) annoValue.getValue();
    }

    private static Boolean getBoolValue(Map<String, AnnotationValue> annoValueMap, String pname, Boolean def) {
        AnnotationValue annoValue = annoValueMap.get(pname);
        if (annoValue == null) return def;
        return (Boolean) annoValue.getValue();
    }

    private static String getEnumConstantName(Map<String, AnnotationValue> annoValueMap, String pname, String def) {
        AnnotationValue annoValue = annoValueMap.get(pname);
        if (annoValue == null) return def;

        VariableElement enumConstant = (VariableElement) getFirstValue(annoValue);
        if (enumConstant == null) return def;
        return enumConstant.getSimpleName().toString();
    }

    private static Object getFirstValue(@Nullable AnnotationValue annoValue) {
        if (annoValue == null) return null;
        Object objValue = annoValue.getValue();
        if (objValue instanceof List<?> list) {
            if (list.isEmpty()) return null;
            annoValue = (AnnotationValue) list.get(0);
            return annoValue.getValue();
        }
        return objValue;
    }
    // endregion
}

File: csharp/Wjybxx.Dson.Apt2/src/AptFieldProps.cs
#region LICENSE

// Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Linq;
using System.Reflection;
using Wjybxx.Commons;
using Wjybxx.Commons.Poet;
using Wjybxx.Dson.Codec.Attributes;
using TypeName = Wjybxx.Commons.Poet.TypeName;

namespace Wjybxx.Dson.Apt2
{
/// <summary>
/// 我们不直接使用<code>Attribute类</code>，
/// 尽可能使两套生成器只在解析层有差异。
/// </summary>
internal class AptFieldProps
{
    public const string DEFAULT_NUMBER_STYLE = "Simple";
    public const string DEFAULT_STRING_STYLE = "Auto";

#nullable disable
    /** 字段序列化时的名字 */
    public string? name;
    /** 取值方法 */
    public string? getter;
    /** 赋值方法 */
    public string? setter;
    /** 是否不可变 */
    public bool isImmutable;

    /** 实现类 -- 会被替换（修正泛型参数） */
    private Type? implType;
    /** 实现类的TypeName缓存 */
    public TypeName? implTypeName;
    /** 写代理方法名 */
    public string? writeProxy;
    /** 读代理方法名 */
    public string? readProxy;

    /** 绑定style -- 需要正确初始化，可能没有注解 */
    public string numberStyle = DEFAULT_NUMBER_STYLE;
    public string stringStyle = DEFAULT_STRING_STYLE;
    public string? objectStyle = null; // 该属性只有显式声明才有效

    /// <summary>
    /// 是否不序列化 -- 非null表示注解指定了值
    /// </summary>
    public bool? ignore;
#nullable enable

    public static AptFieldProps Parse(AptFieldInfo fieldInfo) {
        DsonPropertyAttribute? attribute = fieldInfo.GetAttribute<DsonPropertyAttribute>();
        if (attribute == null) {
            return new AptFieldProps();
        }
        AptFieldProps props = new AptFieldProps();
        props.name = attribute.Name;
        props.getter = attribute.Getter;
        props.setter = attribute.Setter;
        props.isImmutable = attribute.IsImmutable;

        props.numberStyle = EnumUtil.GetName(attribute.NumberStyle)!;
        props.stringStyle = EnumUtil.GetName(attribute.StringStyle)!;
        // objectStyle必须显式声明
        if (attribute.HasObjectStyle) {
            props.objectStyle = EnumUtil.GetName(attribute.ObjectStyle)!;
        }

        props.writeProxy = attribute.WriteProxy;
        props.readProxy = attribute.ReadProxy;
        // 需要将字段的泛型参数拷贝给Impl
        {
            Type? implType = attribute.Impl;
            if (implType != null && implType.IsGenericType) {
                Type[] genericArguments = fieldInfo.FieldType.GetGenericArguments();
                implType = implType.GetGenericTypeDefinition().MakeGenericType(genericArguments);
            }
            if (implType != null) {
                props.implType = implType;
                props.implTypeName = ClassName.Get(implType);
            }
        }
        return props;
    }

    /// <summary>
    /// 解析DsonIgnore注解
    /// </summary>
    /// <param name="fieldInfo"></param>
    public void ParseIgnore(AptFieldInfo fieldInfo) {
        DsonIgnoreAttribute? ignoreAttribute = fieldInfo.GetAttribute<DsonIgnoreAttribute>();
        if (ignoreAttribute == null) {
            return;
        }
        ignore = ignoreAttribute.Value;
    }
}
}